streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala

```diff
@@ -0,0 +1,109 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.scheduler.rate
+
+/**
+ * Implements a proportional-integral-derivative (PID) controller which acts on
+ * the speed of ingestion of elements into Spark Streaming. A PID controller works
+ * by calculating an '''error''' between a measured output and a desired value. In the
+ * case of Spark Streaming the error is the difference between the measured processing
+ * rate (number of elements/processing delay) and the previous rate.
+ *
+ * @see https://en.wikipedia.org/wiki/PID_controller
+ *
+ * @param batchDurationMillis the batch duration, in milliseconds
+ * @param proportional how much the correction should depend on the current
+ *        error. This term usually provides the bulk of correction. A value too large would
+ *        make the controller overshoot the setpoint, while a small value would make the
+ *        controller too insensitive. The default value is -1.
```

Comment:
Assuming that we are switching the signs (i.e. default is 1 instead of -1), what does proportional < 0 mean?
That is just plain wrong! You are increasing the rate when it should be decreased. And anything more than 1 means that you are compensating more than feasible. Consider this. say there is a workload with 1 second batches where the system can process 5 records in 0.8 seconds. Say the rate limit was 10. After first batch with 5 recods in 0.8 seconds process delay (no scheduling delay), the calculation would be (assuming proportional = 1.5)
```
error = rateLimit - numRecords/procDelay 10 - (5/0.8) = 10 - 6.25 = 3.75
newRate = rateLimit - 1.5 * error = 10 - 1.5 * 3.75 = 4.375 
```
4.375 is less than 5 records that the system can handle in 0.8 secs. This is wrong, that is going to happen for proportional > 1. With proportional = 1, the calculation is.
`new rate = 10 - 1.0 * (10 - (5/0.8)) = 6.25  `
This same as it had processed in the previous batch. Perfectly feasible.